### PR TITLE
Notif server

### PR DESCRIPTION
**Pull request by yanethen**
There's a lot of changes, so I closed the previous PR.

Changes:
1. Split the PR into two: (1) Format the code (2) Add the push notification support.
2. Use rate limiting (flask_limiter) on "/token" route to prevent spamming or DDOS attack
3. Add expiration_time field to Token class to help delete outdated tokens from database
4. Load tokens in batches (fireo: fetch and next_fetch); 
- current batch_size = 10 for testing, target 1000?
- the Token.collection.fetch() used in previous code is an generator, will it cause memory issue?
5. For each batch of tokens, group them by first two characters, and process these groups in parallel (concurrent.futures) 
- regarding the previous comment "The other way is to attach a incident Id or timestamp to each token. So for each token/user, we know which notification has been sent, so we just need to send any incidents newer than that timestamp next time."
The incident_publisher publish incidents based on its publisher_status field (Twitter, LinkedIn, push notification), so it looks like there's no need to tag each token for the most recent time the push notification was sent?

### CODE REVIEW

File: firestore/tokens.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+
+class Token(Model):
+    id = IDField(required=True)
+    token = TextField(required=True)
+    expiration_time = DateTime(required=True)
+
+
+def deleteToken(token_id):
+    if Token.collection.delete("token/" + token_id):
```
Comment (lima01) [inline firestore/tokens.py:13]: You can simply do
```
return Token.collection.delete("token/" + token_id)
```

File: firestore/tokens.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        return True
+    return False
+
+
+def registerNewToken(id, token):
+    # return incident id
```
Comment (lima01) [inline firestore/tokens.py:19]: bad comment. pls fix.

File: firestore/tokens.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    expiration_time = DateTime(required=True)
+
+
+def deleteToken(token_id):
+    if Token.collection.delete("token/" + token_id):
+        return True
+    return False
+
+
+def registerNewToken(id, token):
```
Comment (lima01) [inline firestore/tokens.py:18]: let's change it to device_id. also add comment to this method.

Comment (lima01) [inline firestore/tokens.py:18]: Do we always register new token or need to update existing ones too?

File: main.py [hunk 4]
```diff
@@ -26,15 +26,20 @@
...
+from flask_limiter import Limiter
+from flask_limiter.util import get_remote_address
+
 # [END gae_python3_datastore_store_and_fetch_user_times]
 # [END gae_python38_datastore_store_and_fetch_user_times]
 app = Flask(__name__)
 # cors = CORS(app, resources={r"/*": {"origins": "*"}})
 cors = CORS(app)
 firebase_request_adapter = requests.Request()
+limiter = Limiter(get_remote_address, app=app, default_limits=["200/day", "50/hour"])
```
Comment (lima01) [inline main.py:42]: We can commit for now.
But we need to do more research to make sure this is the right way.

Limit on the same remote address might not be appropriate. For example, many people behind the same firewall might share the same public IP. Is there a way to get mobile device id ?

File: main.py [hunk 5]
```diff
@@ -192,7 +210,18 @@ def publish_incidents():
 #     return "success"
 
 
-if __name__ == '__main__':
+@app.route("/token", methods=["PUT"])
+@limiter.limit("100/day;5/hour;1/minute")
+def register_token():
+    deviceId = request.get_json().get("deviceId", None)
+    token = request.get_json().get("token", None)
+    if not token:
```
Comment (lima01) [inline main.py:218]: shall we check device id too?

File: firestore/tokens.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    # return incident id
+    print("New Id:", id)
+    print("New Token:", token)
+    new_token = Token(
+        id=id,
+        token=token,
+        expiration_time=datetime.now() + timedelta(days=30),
```
Comment (lima01) [inline firestore/tokens.py:25]: what does expiration time do?

File: social_media_publishers/notification.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+)
+
+
+class PushNotification(Publisher):
+    def __init__(self) -> None:
+        self.push_client = PushClient()
+
+    def publish(self, incident: Incident) -> datetime:
+        # Fetch tokens in batches
+        batch_size = 10
```
Comment (lima01) [inline social_media_publishers/notification.py:22]: we can do a bigger batch.

Comment (lima01) [reply social_media_publishers/notification.py:22]: Say 100

File: social_media_publishers/notification.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+                # example token: ExponentPushToken[hty9r4BGScZt5gloxXCITM]
+                if group_key not in grouped_tokens:
+                    grouped_tokens[group_key] = []
+                grouped_tokens[group_key].append(token)
+
+            if not grouped_tokens:
+                break
+
+            # Process each group in parallel
+            with concurrent.futures.ThreadPoolExecutor() as executor:
```
Comment (lima01) [inline social_media_publishers/notification.py:43]: pls check Google App engine see if multi-threading is supported or not.

File: firestore/tokens.py (hunk 6 above)
Comment (lima01) [reply firestore/tokens.py:25]: this means if user forgot to use the app for 30 days, we will not send notification anymore. This is not what we want. 

we want to stop sending notifications, only if notification sending failed. I think we can remove the expiration check.

**Review by lima01 — APPROVED** [review]

File: social_media_publishers/notification.py (hunk 8 above)
Comment (yanethen) [reply social_media_publishers/notification.py:43]: multi-threading is not supported on Google App Engine Standard. However, we can use google-cloud Task Queues which is a way of distributing work across many concurrent requests

**Pull request merged**